src/java.base/share/classes/sun/security/ssl/SSLEngineImpl.java

```diff
@@ -789,17 +789,17 @@ public void closeInbound() throws SSLException {
             // Is it ready to close inbound?
             //
             // No exception if the initial handshake is not started.
-            if (!conContext.isInputCloseNotified &&
-                (conContext.isNegotiated ||
-                    conContext.handshakeContext != null)) {
-
-                throw conContext.fatal(Alert.INTERNAL_ERROR,
+            if (!conContext.isInputCloseNotified && (conContext.isNegotiated
+                    || conContext.handshakeContext != null)) {
+                throw new SSLException(
                         "closing inbound before receiving peer's close_notify");
             }
-
-            conContext.closeInbound();
         } finally {
-            engineLock.unlock();
+            try {
+                conContext.closeInbound();
+            } finally {
+                engineLock.unlock();
+            }
```

Comment:
I looked the update further.  It would be nice if the try-statement could support more than one finally blocks in the future so that the code could be more clear.
```
try {
    ...
} finally {
   // the 1st final block
} finally {
   // the 2nd final block
}
```

For the block from 781 to 787, it may be more effective if moving the block out of the synchronization lock (that's, moving 781-787 to line 779.)

Reply:
I'm not following your first comment.  AFAIK, double finally blocks aren't currently an option, unless I'm missing a new language feature.  Or is this just a general "it would be nice if the Java language was able to do..." comment?
```
        try {
            ...
            throw new SSLException(...);
        } finally {
            conContext.closeInbound();
        } finally {
            engineLock.unlock();
        }
```
As to your second question, the model of:
```
method() {
    engineLock.lock();
    try {
        action();
    } finally {
        engineUnlock.unlock();
    }
}
```
is very simple to understand and thus maintain, and is used quite consistently throughout SSLEngineImpl and catches any problems.  IIUC, you are suggesting:
```
    public void closeInbound() throws SSLException {
        if (isInboundDone()) {
            return;
        }

        if (SSLLogger.isOn && SSLLogger.isOn("ssl")) {
            SSLLogger.finest("Closing inbound of SSLEngine");
        }

        engineLock.lock();
        try {
            if (!conContext.isInputCloseNotified && (conContext.isNegotiated
                    || conContext.handshakeContext != null)) {
                throw new SSLException(
                        "closing inbound before receiving peer's close_notify");
            }
        } finally {
            try {
                conContext.closeInbound();
            } finally {
                engineLock.unlock();
            }
        }
    }
```
What is the advantage to changing to this style, other than a very small performance win by not locking in the already closed case?  Isn't the locking code pretty optimized?  SSLLogger might throw a NPE (unlikely), and isInboundDone() just checks a variable, but the code could change down the road.  I'm just not seeing a reason to change the maintainability of this code.

Reply:
> @XueleiFan, I'm not following your first comment.
Sorry for that.  I was wondering something new in the future, which is not doable in the current Java language.  Please just leave it alone.

Reply:
Ah!  Thanks.

Reply:
> IIUC, you are suggesting:
> ... snipped ...
Yes.

> What is the advantage to changing to this style, other than a very small performance win by not locking in the already closed case?

The advantage is mainly about the performance, as I/O of SSLLogger could be slow.  The SSLSocketImpl code does not put the similar block locked.  But it may be not the purpose of this update. I'm fine if you want to leave it alone.

Reply:
I can see your point for these two methods (`close*bound()`), but we are using `SSLLogger` behind the `engineLock` throughout the implementation.  I think I will leave as is, rather than try to squeak this little bit out when logging is on and active.